main.ts

```diff
@@ -168,69 +154,64 @@ export default class FilenameHeadingSyncPlugin extends Plugin {
       return;
     }
 
-    const editor = view.editor;
-    const doc = editor.getDoc();
-    const cursor = doc.getCursor();
-
-    const docstart = this.findDocstart(doc);
-
-    const foundHeading = this.findHeading(doc, docstart);
     const sanitizedHeading = this.sanitizeHeading(file.basename);
-
-    if (foundHeading !== null) {
-      if (this.sanitizeHeading(foundHeading.Text) !== sanitizedHeading) {
-        this.replaceLine(doc, foundHeading.LineNumber, `# ${sanitizedHeading}`);
-        doc.setCursor(cursor);
-      }
-      return;
-    }
-
-    if (docstart > 0) {
-      this.insertLine(doc, docstart + 1, `\n# ${sanitizedHeading}`);
-    } else {
-      this.insertLine(doc, docstart, `# ${sanitizedHeading}`);
-    }
-    doc.setCursor(cursor);
+    this.app.vault.read(file).then((data) => {
+      const lines = data.split('\n');
+      const start = this.findNoteStart(lines);
+      const heading = this.findHeading(lines, start);
+
+      if (heading !== null) {
+        if (this.sanitizeHeading(heading.text) !== sanitizedHeading) {
+          this.replaceLine(
```

Comment:
Is replaceLine alone here enough? Does `data` need to get written back manually, or is obsidian handling that already?

Reply:
The `data` needs to be written back manually. I decided to perform the writing in the replaceLine function, to keep our sync function from becoming messy. So, please check out the replaceLine function, it's essentially completely transformed.

Reply:
Ah okay now I see it! How about `replaceLineInFile`? That makes it more clear what's actually happening. 

Personally also not a fan of the `insertInstead` parameter. Semantically `replaceLine` means *replace*, so having a function that's called `replace` doing an insertion instead is a bit odd. Can we split it into insert/replace?

Reply:
Yeah, I agree on adding the `InFile`, that makes it much clearer.

I also agree the `insertInstead` parameter is a bit hacky and we could of course split the function in two (insert/replace). However, that way we are duplicating a bit of code, which can also be considered a 'bad practice' and makes it more complicated to modify later on. 

If you are fine with that, then I'll make the change, but can we maybe find a word that serves the purpose of both operations? All I can think of is `editLineInFile` but inserting a line isn't really editing a line, is it? Should we go for a long `addOrEditLineInFile`? 😛

There's also the word 'upsert' used for update/insert in database management, but I'm not sure how I feel about it.

Reply:
Thought about this a bit and couldn't come up with a better word either. I think it's fine to split and have a bit of duplication, it also makes the functions itself a bit slimmer.